### PR TITLE
Validate target range [0, 1] in binary_cross_entropy_with_logits

### LINKED ISSUE #110865: BCEWithLogitsLoss: Check if labels within zero and one
### 🚀 The feature, motivation and pitch

`BCEWithLogitsLoss` currently performs no kind of checking wrt. the targets whatsoever. This means, that calling the loss function with targets not contained within the interval [0,1] succesfully computes values without throwing an error. However, these values don't have any kind of meaning.

In practice, though, passing invalid targets to `BCEWithLogitsLoss` can easily happen and can go unnoticed! Such mistakes can render whole experiments invalid; **if** they are even noticed in the first place!

For instance, this can happen if multi-class labels and predictions (i.e. 0,1,2,...,n) are erroneously used with BCEWithLogitsLoss. *Shouldn't this result in a shape-mismatch due to the prediction being a categorical distribution?* No, both predictions and targets will be broadcastable to each other in that case. In fact, commonly, their shape will be identical.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rajfirke**
Fixes #110865

`BCELoss` validates that target values are in `[0, 1]` (per-element TORCH_CHECK on CPU, CUDA_KERNEL_ASSERT on CUDA), but `BCEWithLogitsLoss` does not, silently accepting any target value and producing meaningless loss values. As @mikaylagawarecki noted: "Since we validate target for BCELoss it seems reasonable to add the same validation for BCEWithLogitsLoss."

```python
loss = nn.BCEWithLogitsLoss()
loss(torch.randn(5), torch.tensor([0., 1., 2., -1., 0.5]))  # silently succeeds
```

**Fix** across both eager and torch.compile paths:
- `aten/src/ATen/native/Loss.cpp`: Add `TORCH_CHECK` at the entry of `binary_cross_entropy_with_logits` (single dispatch point for all devices via CompositeExplicitAutograd)
- `torch/_decomp/decompositions.py`: Add `torch._check` in the decomposition for the torch.compile path
- `test/test_nn.py`: Add `test_bce_with_logits_target_range` following the existing `test_bce_loss_input_range` pattern

Two prior PRs (#177506, #113433) attempted this fix but went stale without merge.

cc @mikaylagawarecki @jbschlosser @albanD

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -347,6 +347,9 @@ Tensor& binary_cross_entropy_backward_out_cpu(const Tensor& grad, const Tensor&
 }
 
 Tensor binary_cross_entropy_with_logits(const Tensor& input, const Tensor& target, const std::optional<Tensor>& weight_opt, const std::optional<Tensor>& pos_weight_opt, int64_t reduction) {
+  TORCH_CHECK(
+      target.numel() == 0 || (at::all(target >= 0).item<bool>() && at::all(target <= 1).item<bool>()),
```
Comment (Skylion007) [inline aten/src/ATen/native/Loss.cpp]: Why not just at::minmax range check here

Comment (rajfirke) [reply aten/src/ATen/native/Loss.cpp]: Good call — updated to use `at::aminmax` for a single-pass min/max check instead of two separate `at::all` scans. Thank you.

**Review by soulitzer — CHANGES_REQUESTED** [review]
Doesn't this cause a sync?

**Comment by rajfirke** [comment]
Good point @soulitzer — `at::aminmax(target).item<double>()` would indeed cause a host sync on CUDA, which is a performance concern.

For reference, `binary_cross_entropy` (without logits) avoids this by checking per-element inside the CPU/CUDA kernels, but `binary_cross_entropy_with_logits` is `CompositeExplicitAutograd` with no per-element kernel.

Options:
1. Only validate on CPU tensors (skip for CUDA) — matches BCE's approach where CUDA uses `CUDA_KERNEL_ASSERT` (no host sync)
2. Keep the sync — arguably acceptable since incorrect targets are a bug in user code
3. Remove the C++ check and only validate in the Python wrapper (`F.binary_cross_entropy_with_logits`) with a device guard

Which approach would you prefer? Happy to adjust.

**Comment by rajfirke** [comment]
Gentle ping @soulitzer — I acknowledged the host sync concern above and noted that `binary_cross_entropy` (without logits) already performs a similar check. Would you prefer a lazy/deferred check approach, or should we drop the eager validation? Happy to adjust.

**Comment by rajfirke** [comment]
Gentle ping — target range `[0, 1]` validation for `binary_cross_entropy_with_logits` is ready for review. cc @soulitzer @Skylion007
